spec/20-AMQP_FORMAT.md

```diff
@@ -1,3 +1,132 @@
+From the distributed tracing and telemetry correlation perspective it is a
+known problem to be able to correlate a component that placed message and
+component that processed it later. This specification describes how trace
+context MUST be encoded into AMQP messages.
+
+## Trace context fields placement in a message
+
+AMQP defines message as a payload with the additional annotations sections.
+There are two annotation sections this specification refers -
+"application-properties" and "message-annotations". See
+[3.2](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#section-message-format)
+of *OASIS Advanced Message Queuing Protocol (AMQP) Version 1.0, Part 3:
+Messaging* (AMQP specification).
+
+AMQP message section "application-properties" is immutable collection of
+properties that is defined by message publisher and can be read by the message
+consumer. Message brokers cannot mutate those properties. Section
+"message-annotations" is designed for message brokers to use and can be mutated
+during the message processing.
+
+Fields `traceparent` and `tracestate` SHOULD be added to the message in the
+application-properties section by message publisher. Once the message has been
+created, it’s no longer permissible to edit the bare message. So if it were
+necessary to annotate the message inside the middleware as it transits, that
+MUST happen in the “message-annotations” section, using the
+“application-properties” as a base. Message reader SHOULD construct the full
+trace context by reading `traceparent` and `tracestate` fields from the
+“message-annotations” first and if not exist - from “application-properties”.
+
+## Trace context and failed read/write operations
+
+This specification defines how to propagate context from publisher thru broker
+to ultimate reader. It does not define any additional transport level correlation
+constructs that can be used to investigate failed publish or read operations.
+
+It is recommended, however, for AMQP implementations to make the best effort
+attempt to read the trace context from the message and use it while reporting
+such problems.
+
+TODO: do we need to define a way to specify trace context of read operation? Is
+there anything in AMQP protocol that can be used to carry this context?
+
+## `traceparent` AMQP format
+
+The field `traceparent` MUST be encoded and decoded using [binary
+protocol](..\extension-binary.html) and stored as a binary type defined in
+section
+[1.6.19](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-binary)
+of AMQP specification.
+
+Property name MUST be `traceparent` - all lowercase without delimiters.
+
+## `tracestate` AMQP format
+
+The field `tracestate` MUST be encoded and decoded as a string to string map.
+See definition of type map in section
+[1.6.23](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-map)
+of AMQP specification.
```

Comment:
Comment from @calohmn

Considering the case that the `traceparent` and `tracestate` fields are transported via _application-properties_:

About application-properties, the AMQP spec says (http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-application-properties):

> The keys of this map are restricted to be of type string (which excludes the possibility of a null key) and the values are restricted to be of simple types only, that is, excluding map, list, and array types.

This looks incompatible with wanting to store a `tracestate` AMQP map in there.

Reply:
HTTP headers are constrained to strings and you still manage to put a map there.

Reply:
I was referring to the AMQP `map` datatype being mentioned for the `tracestate` field value here. 
As for using a string value instead, I guess it should be clarified then, how exactly the string-to-string map is to be encoded in that string value.

Reply:
So it's not possible to put a map into application-properties. 
And even if it was, it is impractical to ask everyone to parse tracestate - the majority of services will blindly propagate it:
1. receive incoming http request with tracestate as string
2. put tracestate into EventHub message. If it is map will require parsing it in the first place which should not be necessary or desired by this service

Reply:
it is impractical to ask everyone to parse tracestate - the majority of services will blindly propagate it:

1. receive incoming http request with tracestate as string
2. put tracestate into EventHub message. If it is a map - it will require parsing it in the first place which should not be necessary for arbitrary service

So my proposal: it should be either a string (exactly like http-header) or byte array to save n message size (and this spec should define how it is formatted).